uses message type instead of topic name

bsubei opened the pull request:
Instead of using the topic name for choosing messages to compress/decompress, it checks whether the message types are `sensor_msgs/Image` or `sensor_msgs/CompressedImage`.

I also had to modify the `image_topic_basename()` function to return None instead of raise an exception. In the case that it can't find a basename, the script just uses the old topic name and compresses/decompresses if the message types match.

I ran this through the unittest in the package, and it checked out out fine. I also ran it manually, and it seemed ok. Let me know if it's too hacky, maybe I can think of something better.

tfoote commented:
I like the approach of using the types. The challenge I see is that if you compress and uncompress a bag it must get back to the same state. I think with the logic you have at the moment somthing like this would happen. 

`/foo (sensor_msgs/Image)` >> compression >> `/compressed` >> decompression >>`/image_raw`

Which will break anything relying on the topic `/foo` The unit tests don't cover this case as it's currently not supported, but if trying to generalize the unit tests would need to also be generalized. The same potential issue stands for non-standard compressed image topics. The current designed to support only the standard image_transport interface.

bsubei commented:
Sorry, I'm a little confused as to what the desired behavior should be.

Here's what it currently does to any bags with image topic names that don't end with `/image_raw`. I use an example topic name of `/rgb` to start with:
`/rgb (sensor_msgs/Image)` >> compression >> `/rgb/compressed (sensor_msgs/CompressedImage)` (also publishes encoding topic `/rgb/encoding`) >> decompression >> `/rgb/image_raw (sensor_msgs/Image)`

Is that the desired behavior?

I can take a look at the current unittests and see if I can extend them to cover these new cases.

tfoote commented:
No that's not the desired behavior, but I think will be the behavior of the current pull-request. If it starts as `/rgb (sensor_msgs/Image)` it should end as `/rgb (sensor_msgs/Image)` 

The bag should be the same if compressed then decompressed.

bsubei commented:
I see... The simplest fix I can see is to send some sort of flag over on the `encoding` topic to indicate that the topic names are non-standard. Is that acceptable or too hacky?

bsubei commented:
the new fix results in the desired behavior, but I had to "borrow" the `encoding` topic (only when a non-standard topic name is compressed) to signal non-standard topic name. When it decompresses, it returns the `encoding` string back to normal. I feel like it's too hacky, but I couldn't think of anything better.

tfoote commented:
Overall that looks good. Could you change `non-default` to be `no-basename` to be clearer about which behavior it's relevant to. And squash the commits. 

The encoding topic is already injected just to capture this extra metadata. A cleaner solution would be to encode the data in the string with markup, but I think comma separated list seems reasonable for now. I don't know of another value.

bsubei committed:
non-standard topic names decompressed back to original topic name.

An extra string flag (no-basename) in the encoding.data string is passed into the encoding topic, which signifies to the decompressor that the topic name has no basename. It then uses the original topic name instead of appending image_raw. The encoding string is then returned to normal. This does not affect any of the standard topic names (ending with image_raw etc.)
Some unittests were extended to check the encoding string based on topic names.

bsubei committed:
Merge remote-tracking branch 'upstream/indigo-devel' into indigo-devel

bsubei commented:
oh, wow (I'm terrible at picking names). I squashed the commits, and I also included a simple test. It ran successfully, so I don't think I broke anything there.

tfoote commented on test/test_rosbag_image_compressor.py:
```diff
@@ -107,8 +108,23 @@ def test_bag_unbag(self):
+            with rosbag.Bag(intermediate, 'r') as bag:
+                for topic, msg, t in bag:
+                    # non-standard topic name should have flag in encoding
+                    if topic == '/foo':
```
Shouldn't this be `/foo/encoding`?

bsubei replied:
I think you're correct. It works with either topic `foo` and then checking `msg.encoding.endswith()` or using topic `foo/encoding` and then checking `msg.endswith()`... I'll change this

tfoote commented on test/test_rosbag_image_compressor.py:
```diff
@@ -107,8 +108,23 @@ def test_bag_unbag(self):
+                    if topic == '/foo':
+                        self.assertTrue(msg.encoding.endswith(', no-basename'))
+                    # standard topic name should not have flag in encoding
+                    elif topic == 'foo/bar/image_raw':
```
Shouldn't this be `/foo/bar/encoding`?

bsubei replied:
For some reason, this condition is never encountered in the tests, although the TestBag writes to both topics...

bsubei replied:
nevermind, I forgot the first slash for the topic name. I will fix these

bsubei committed:
fixes incorrect topic names in test, also checks final output encoding.
